src/function/numeric_functions.cpp

```diff
@@ -0,0 +1,439 @@
+    }
+    case 'n':
+    case 'N': {
+      static constexpr char kNo[] = "no";
+      if (strncasecmp(start, kNo, trimmed_len) == 0) {
+        return false;
+      }
+      break;
+    }
+    case 'o':
+    case 'O': {
+      // 'o' not enough to distinguish between on/off
+      static constexpr char kOff[] = "off";
+      static constexpr char kOn[] = "on";
+      if (strncasecmp(start, kOff, (trimmed_len > 3 ? trimmed_len : 3)) == 0) {
+        return false;
+      } else if (strncasecmp(start, kOn, (trimmed_len > 2 ? trimmed_len : 2)) ==
+                 0) {
+        return true;
+      }
+      break;
+    }
+    case '0': {
+      if (trimmed_len == 1) {
+        return false;
+      } else {
+        return true;
+      }
+    }
+    case '1': {
+      if (trimmed_len == 1) {
+        return true;
+      } else {
+        return false;
+      }
+    }
+    default: { break; }
+  }
+
+  // Error
+  codegen::RuntimeFunctions::ThrowInvalidInputStringException();
+  __builtin_unreachable();
+}
+
+int8_t NumericFunctions::InputTinyInt(
+    UNUSED_ATTRIBUTE const codegen::type::Type &type, const char *ptr,
+    uint32_t len) {
+  PELOTON_ASSERT(ptr != nullptr && "Input is assumed to be non-NULL");
+  return ParseInteger<int8_t>(ptr, len);
+}
+
+int16_t NumericFunctions::InputSmallInt(
+    UNUSED_ATTRIBUTE const codegen::type::Type &type, const char *ptr,
+    uint32_t len) {
+  PELOTON_ASSERT(ptr != nullptr && "Input is assumed to be non-NULL");
+  return ParseInteger<int16_t>(ptr, len);
+}
+
+int32_t NumericFunctions::InputInteger(
+    UNUSED_ATTRIBUTE const codegen::type::Type &type, const char *ptr,
+    uint32_t len) {
+  PELOTON_ASSERT(ptr != nullptr && "Input is assumed to be non-NULL");
+  return ParseInteger<int32_t>(ptr, len);
+}
+
+int64_t NumericFunctions::InputBigInt(
+    UNUSED_ATTRIBUTE const codegen::type::Type &type, const char *ptr,
+    uint32_t len) {
+  PELOTON_ASSERT(ptr != nullptr && "Input is assumed to be non-NULL");
+  return ParseInteger<int64_t>(ptr, len);
+}
+
+double NumericFunctions::InputDecimal(
+    UNUSED_ATTRIBUTE const codegen::type::Type &type, const char *ptr,
+    uint32_t len) {
+  PELOTON_ASSERT(ptr != nullptr && "Input is assumed to be non-NULL");
+  if (len == 0) {
+    codegen::RuntimeFunctions::ThrowInvalidInputStringException();
+    __builtin_unreachable();
+  }
+
+  const char *start = ptr;
+  const char *end = ptr + len;
+
+  // We don't trim because std::strtod() does the trimming for us
+
+  // TODO(pmenon): Optimize me later
+  char *consumed_ptr = nullptr;
+  double ret = std::strtod(ptr, &consumed_ptr);
+
+  if (unlikely_branch(consumed_ptr == start)) {
+    if (errno == ERANGE) {
+      codegen::RuntimeFunctions::ThrowOverflowException();
+      __builtin_unreachable();
+    } else {
+      codegen::RuntimeFunctions::ThrowInvalidInputStringException();
+      __builtin_unreachable();
+    }
+  }
+
+  // Eat the rest
+  while (consumed_ptr < end && *consumed_ptr == ' ') {
```

Comment:
Does this mean, we basically only allow spaces after the actual parsed decimal?

Reply:
We allow spaces both before and after actually. `std::strtod` handles spaces beforehand. We handle trailing spaces.